### ISSUE TITLE
Incorrect results when stack multiple CNN layers

### ISSUE THREAD

**Issue by lzhengchun** [intel/mkl-dnn#566]
```cpp
removed code that has bug
```

Based on the inference_f32.cpp example, I wrap up a function for conv2d because my model has 16 conv layers.
I **push_back** input tensor (NCHW) to the **net_args** before calling the above function to build conv2d layers. all conv2d weights are set to 1 for testing.

With above code, it gives me correct results only when there is only one conv2d layer. 

when there is more than one layer, the final output will be *inf*, it seems that some memory are not initialized.

but if I do not use that **tag::any** for either **conv_src_md** or **conv_dst_md**, it gives me correct results (runs much slower). I have spent a day on this but still do not understand what was wrong. 

One thing really confused me is the **reorder**, should it **execute** when I build the graph or should it **execute** with the entire graph?  I see examples used both ways, as the code I commented. 

**memory_format_propagation.cpp** explicitly sees to **execute** the **reorder** while building the graph (and there is about s.wait()), I guess at that time the source memory of **reorder**  do not have actually value, it should not give correct **reorder** results I thought.  I tried both, nether gives me correct results. 

help wanted!

Thanks

**Comment by emfomenk** [comment]
Hi @lzhengchun,

Could you please make a gist with the whole example that we can build and run? I look through the code and don't see any issues.

> One thing really confused me is the reorder, should it execute when I build the graph or should it execute with the entire graph? I see examples used both ways, as the code I commented.

Reorders typically are part of the graph itself. They must be part of the graph whenever this is the data that flows (e.g. dst of one convolution being reorder to src of another convolution with the different format). However, if we talk about inference, you can reorder the weights in the appropriate format in advance. It is perfectly fine if they would be part of the graph as well, but that might slightly increase the inference time.

**Comment by lzhengchun** [comment]
Hi @emfomenk Thanks for your response. One more question about reorder, for inference how can I know the best layout format for a given device if I want to reorder the weights in advance for better performance? Thanks!

I paste my source here, you should be able to compile and run. it gives my correct results with **format_tag::any** (line 46) when there is only one layer (i.e., comment line 135). it also gives to correct results when there is two layers but use **format_tag::nchw** in line 46 (but runs much slower).

```cpp
#include <assert.h>
#include <chrono>
#include <numeric>
#include <vector>
#include <unordered_map>
#include <iostream>
#include <fstream>

#include "tomogan.hpp"

using namespace std;
using namespace dnnl;

void conv2d_opt( 
    memory::dim knl_sz,
    memory::dim n_knl,
    std::vector<primitive> &net,
    std::vector<std::unordered_map<int, memory>> &net_args,
    engine &eng, 
    stream &s,
    bool has_relu){
    
    auto in_dims = net_args.back()[DNNL_ARG_DST].get_desc().data.dims;
    memory::dim batch    = in_dims[0];
    memory::dim in_img_c = in_dims[1];
    memory::dim in_img_h = in_dims[2];
    memory::dim in_img_w = in_dims[3];

    memory::dims conv_src_tz = {batch, in_img_c, in_img_h, in_img_w};
    memory::dims conv_weights_tz = {n_knl, in_img_c, knl_sz, knl_sz};
    memory::dims conv_bias_tz = {n_knl};
    memory::dims conv_dst_tz = {batch, n_knl, in_img_h, in_img_w};
    memory::dims conv_strides = {1, 1};
    memory::dims conv_padding = {knl_sz/2, knl_sz/2}; // padding as the same

    /// Create memory that describes data layout in the buffers. here we use
    /// tag::oihw for weights.
    auto user_weights_memory   = memory({{conv_weights_tz}, memory::data_type::f32, memory::format_tag::oihw}, eng);
    auto conv_user_bias_memory = memory({{conv_bias_tz},    memory::data_type::f32, memory::format_tag::x},    eng);
    
    /// Create memory descriptors with layout tag::any. The `any` format enables
    /// the convolution primitive to choose the data format that will result in
    /// best performance based on its input parameters (convolution kernel
    /// sizes, strides, padding, and so on). If the resulting format is different
    /// from `nchw`, the user data must be transformed to the format required for the convolution.
    auto conv_src_md     = memory::desc({conv_src_tz},     memory::data_type::f32, memory::format_tag::any);
    auto conv_bias_md    = memory::desc({conv_bias_tz},    memory::data_type::f32, memory::format_tag::x);
    auto conv_weights_md = memory::desc({conv_weights_tz}, memory::data_type::f32, memory::format_tag::any);
    auto conv_dst_md     = memory::desc({conv_dst_tz},     memory::data_type::f32, memory::format_tag::any);

    auto user_src_memory = net_args.back()[DNNL_ARG_DST];
    /// Create a convolution descriptor
    auto conv_desc = convolution_forward::desc(prop_kind::forward_inference,
            algorithm::convolution_auto, conv_src_md, conv_weights_md,
            conv_bias_md, conv_dst_md, conv_strides, conv_padding, conv_padding);

    /// Create a convolution primitive descriptor. Once created, this
    /// descriptor has specific formats instead of the 'any' format specified
    /// in the convolution descriptor.
    auto conv_prim_desc = convolution_forward::primitive_desc(conv_desc, eng);
    
    bool need_reorder_src = conv_prim_desc.src_desc() != user_src_memory.get_desc();
    auto conv_src_memory = need_reorder_src ? memory(conv_prim_desc.src_desc(), eng) : user_src_memory;
    if (need_reorder_src) {
        std::cout << "!!!!! src mem needs reorder" << std::endl;
        // reorder(user_src_memory, conv_src_memory).execute(s, 
        //         {{DNNL_ARG_FROM, user_src_memory}, {DNNL_ARG_TO, conv_src_memory}});
        // s.wait();
        net.push_back(reorder(user_src_memory, conv_src_memory));
        net_args.push_back({{DNNL_ARG_FROM, user_src_memory},
                            {DNNL_ARG_TO, conv_src_memory}});
    }

    bool need_reorder_weights = conv_prim_desc.weights_desc() != user_weights_memory.get_desc();
    auto conv_weights_memory  = need_reorder_weights ? memory(conv_prim_desc.weights_desc(), eng) : user_weights_memory;
    if (need_reorder_weights) {
        std::cout << "!!!!! weight mem needs reorder" << std::endl;
        // reorder(user_weights_memory, conv_weights_memory).execute(s, 
        //     {{DNNL_ARG_FROM, user_weights_memory}, {DNNL_ARG_TO, conv_weights_memory}});
        // s.wait();
        net.push_back(reorder(user_weights_memory, conv_weights_memory));
        net_args.push_back({{DNNL_ARG_FROM, user_weights_memory},
                            {DNNL_ARG_TO, conv_weights_memory}});
    }

    /// Create a memory primitive for output.
    auto conv_dst_memory = memory(conv_prim_desc.dst_desc(), eng);

    /// Create a convolution primitive and add it to the net.
    net.push_back(convolution_forward(conv_prim_desc));
    net_args.push_back({{DNNL_ARG_SRC,     conv_src_memory},
                        {DNNL_ARG_WEIGHTS, conv_weights_memory},
                        {DNNL_ARG_BIAS,    conv_user_bias_memory},
                        {DNNL_ARG_DST,     conv_dst_memory}});
    // relu
    if(has_relu){
        /// Create the relu primitive. For better performance, keep the input data
        /// format for ReLU (as well as for other operation primitives until another
        /// convolution or inner product is encountered) the same as the one chosen
        /// for convolution. Also note that ReLU is done in-place by using conv memory.
        auto relu_desc = eltwise_forward::desc(prop_kind::forward_inference,
                algorithm::eltwise_relu, net_args.back()[DNNL_ARG_DST].get_desc(), 1.0f);
        auto relu_prim_desc = eltwise_forward::primitive_desc(relu_desc, eng);

        net.push_back(eltwise_forward(relu_prim_desc));
        net_args.push_back({{DNNL_ARG_SRC, net_args.back()[DNNL_ARG_DST]},
                            {DNNL_ARG_DST, net_args.back()[DNNL_ARG_DST]}});
    }
}

int main(int argc, char const *argv[]){
    /// Initialize an engine and stream. 
    dnnl::engine eng(dnnl::engine::kind::cpu, 0);
    stream s(eng);
    const memory::dim batch = 1;
    const memory::dim img_sz = 1024;
    const memory::dim img_ch = 3;
    /// Create a vector for the primitives and a vector to hold memory
    /// that will be used as arguments.
    std::vector<primitive> net;
    std::vector<std::unordered_map<int, memory>> net_args;

    memory::dims conv_src_tz = {batch, img_ch, img_sz, img_sz};
    /// Create memory that describes data layout in the buffers. here we use
    /// tag::nchw for input data.
    auto user_src_memory = memory({{conv_src_tz}, memory::data_type::f32, memory::format_tag::nchw}, eng);
    /// Allocate buffers for input and output data, weights, and bias.
    std::vector<float> user_src(batch * img_ch * img_sz * img_sz);
    // copy input image to input buffer and then write to engine memory
    write_to_dnnl_memory(user_src.data(), user_src_memory);
    net_args.push_back({{DNNL_ARG_SRC, user_src_memory},
                        {DNNL_ARG_DST, user_src_memory}});

    conv2d_opt(3, 8, net, net_args, eng, s, true);
    conv2d_opt(3, 32, net, net_args, eng, s, true);
    
    for (auto args : net_args){
        if(args.find(DNNL_ARG_SRC) != args.end()){
            auto in_dims = args[DNNL_ARG_SRC].get_desc().data.dims;
            printf("Input: %d x %3d x %d x %d", in_dims[0], in_dims[1], in_dims[2], in_dims[3]);
        }else{
            auto in_dims = args[DNNL_ARG_MULTIPLE_SRC + 0].get_desc().data.dims;
            printf("Input: %d x %3d x %d x %d, ", in_dims[0], in_dims[1], in_dims[2], in_dims[3]);
            in_dims = args[DNNL_ARG_MULTIPLE_SRC + 1].get_desc().data.dims;
            printf("%d x %3d x %d x %d", in_dims[0], in_dims[1], in_dims[2], in_dims[3]);
        }

        auto out_dims = args[DNNL_ARG_DST].get_desc().data.dims;
        printf(" => Output: %d x %3d x %d x %d\n", out_dims[0], out_dims[1], out_dims[2], out_dims[3]);
    }

    // set all weights to 1 and bias to zero for debug
    for(int i = 0; i < net_args.size(); i++){
        if(net_args[i].find(DNNL_ARG_WEIGHTS) == net_args[i].end()){
            continue;
        }

        auto in_dims = net_args[i][DNNL_ARG_WEIGHTS].get_desc().data.dims;
        unsigned int wbuf_size = in_dims[0] * in_dims[1] * in_dims[2] * in_dims[3];

        std::vector<float> conv_weights(wbuf_size);
        for(auto i = 0; i < conv_weights.size(); i++){
            conv_weights[i] = 1;
        }
        
        std::vector<float> conv_bias(in_dims[0]);
        for(auto i = 0; i < conv_bias.size(); i++){
            conv_bias[i] = 0;
        }
        
        printf("conv layer %d Weights: %d x %d x %d x %d\n", i, in_dims[0], in_dims[1], in_dims[2], in_dims[3]);
        write_to_dnnl_memory(conv_weights.data(), net_args[i][DNNL_ARG_WEIGHTS]);
        write_to_dnnl_memory(conv_bias.data(),    net_args[i][DNNL_ARG_BIAS]);
    }
    // explicitly set input value for debug
    auto in_dims = net_args[0][DNNL_ARG_DST].get_desc().data.dims;
    unsigned int in_buf_size = in_dims[0] * in_dims[1] * in_dims[2] * in_dims[3];
    std::vector<float> input_buf(in_buf_size);
    for(auto n = 0; n < in_dims[0]; n++)
        for(auto c = 0; c < in_dims[1]; c++)
            for(auto h = 0; h < in_dims[2]; h++)
                for(auto w = 0; w < in_dims[3]; w++){
                    input_buf[n * in_dims[1] * in_dims[2] * in_dims[3] +\
                              c * in_dims[2] * in_dims[3] + \
                              h * in_dims[3] + w] = c;
                }

    write_to_dnnl_memory(input_buf.data(), net_args[0][DNNL_ARG_DST]);

    auto mdl_exe_st = chrono::steady_clock::now();

    assert((net.size()+1) == net_args.size() && "something is missing");
    for (size_t i = 0; i < net.size(); ++i)
        net.at(i).execute(s, net_args.at(i+1));
    s.wait();
    
    auto mdl_exe_ed = chrono::steady_clock::now();
    printf("It takes %.3f ms to execute the model!\n", \
           chrono::duration_cast<chrono::microseconds>(mdl_exe_ed - mdl_exe_st).count()/1000.);

    // <should check if reorder needed> copy results from dnnl device to cpu mempry
    auto output_dims = net_args.back()[DNNL_ARG_DST].get_desc().data.dims;
    auto output_size = output_dims[0] * output_dims[1] * output_dims[2] * output_dims[3];
    auto output_buf = new float[output_size]();
    read_from_dnnl_memory(output_buf, net_args.back()[DNNL_ARG_DST]);

    double sum = 0;
    for(size_t i = 0; i < output_size; i++){
        sum += output_buf[i];
    }
    printf("results checksum: %lf\n", sum);

    delete[] output_buf;
}
```

**Comment by emfomenk** [comment]
Hi @lzhengchun,

The problem seems to happen here:
``` cpp
    // set all weights to 1 and bias to zero for debug
    for (int i = 0; i < net_args.size(); i++){
        if(net_args[i].find(DNNL_ARG_WEIGHTS) == net_args[i].end()){
            continue;
        }
        ...
    }
```

Here, what happens is that you initialize the buffers with `1.f` for **already reordered** weights. So before you start the program you have the following picture:
```
user_src(nchw)     : initialized
user_weights(oihw) : uninitialized      !!! <-- should be initialized instead of conv_weights
conv_weights(any)  : initialized with 1 !!! <-- could be left uninitialized
...
```
But then at run time there is a reorder: `user_weights --> conv_weights`, which overrides the ones you put with whatever garbage was in uninitialized user_weights. Please try change the program so that it initialize the original weights, then reorder will put `1.f` appropriately to the conv_weights.

Also, I am not sure I understand why the code works for the case where there is only one convolution. To me the problem should appear whenever the weights reorder is called (and that should happen for one layer as well). Moreover, when I tried to reproduce the problem on my side, I got incorrect results for one layer too. So it is really strange that for you it works. Well, anyways, the program should be fixed as I mentioned above.

Few notes:
1. The program works for `format::tag = nchw` because in this case the weights would be in `oihw` format, and no reorder is required. Hence, the results are correct.
2. I see you want to run conv + relu. In the code you share with us ReLU is a separate primitive. Please note that it is beneficial to **fuse** it with the convolution computations. [Here](https://intel.github.io/mkl-dnn/dev_guide_attributes_post_ops.html) is the relevant part in the documentation. Also, there is an example that shows the benefits of fusing: [performance_profiling_cpp](https://intel.github.io/mkl-dnn/performance_profiling_cpp.html)
I am not sure I understa

**Comment by lzhengchun** [comment]
@emfomenk Thanks very much!! bug fixed.
FYI One layer works because no **reorder** needed for my device (Gen9 GPU). i.e., the preset nchw for input matches with oihw. I did not see any reorder logs when run with only one layer. 
Because the output of the 1st is no longer nchw, the weights of the 2nd layer and after need **reorder**. 
BTW, how to know which layout was chosen by the primitive? i.e., how to print **conv_prim_desc.weights_desc()** ?

Thanks

**Comment by emfomenk** [comment]
> BTW, how to know which layout was chosen by the primitive? i.e., how to print conv_prim_desc.weights_desc() ?

Not an easy task :)

In addition to the standard native layouts like `NCHW`, `NHWC`, `OIHW`, `HWIO`, etc... DNNL supports a variety of other formats (most of them are called blocked). The blocked formats typically block one or several dimensions in... well... blocks. This allows us to vectorize the computations to achieve the maximum performance. For instance, on CPU with AVX512 instructions we use `nChw16c` format for the activations. This allows us to load 16 channels into one `zmm` register at once.
The blocked format is described with the following structure [`dnnl_blocked_desc_t`](https://intel.github.io/mkl-dnn/structdnnl__blocking__desc__t.html) which is a part of [`dnnl_memory_dest_t`](https://intel.github.io/mkl-dnn/structdnnl__memory__desc__t.html).
To get better understanding of the memory formats I recommend to go through [Understanding Memory Formats](https://intel.github.io/mkl-dnn/dev_guide_understanding_memory_formats.html) article.

So coming back to the question on how to know the layout. The expectation is that users don't need to know what exactly the layout is. it is sufficient for them just to know whether the layout of the tensor is the same they currently have or not.
Consider the situation:
- You have pre-trained weights in some well known format, say `HWIO` from TensorFlow.
- You create a convolution with weights format `any` to allow DNNL to choose the best memory format.
- Now, the only question is whether you need to reorder the weights you have to some peculiar memory format, or it happens that DNNL wants exactly `HWIO` memory format.

For that we have a comparison operator ([`==`](https://github.com/intel/mkl-dnn/blob/master/include/dnnl.hpp#L1473), [`!=`](https://github.com/intel/mkl-dnn/blob/master/include/dnnl.hpp#L1477)) between memory descriptors.

To summarize:

| Use-case | Regular | Advanced |
| :-- | :-- | :-- |
| Way to understand memory format | Only by comparing with other memory formats, typically native formats like `NCHW` or `HWIO` | By inspecting the `dnnl_memory_desc_t` structure | 
| Example | Typical integration of DNNL to frameworks or other users' programs | For developers who want to have a custom operation that directly works on DNNL memory format (i.e. w/o reordering it back to the native layout) |

**Comment by lzhengchun** [comment]
Thanks @emfomenk Understood. I have pre-trained weights and wants to avoid **reorder**, so I was thinking if I know which layout the primitive prefers, I can reorder my weights to that layout before loading.  so that DNNL doesnot need to reorder for each inference. 

So, it seems impossible to avoid reorder, right? I have tried several standard layouts like OIHW, IOHW etc. All need reorder based on the (==, !=) operation.

**Comment by rsdubtso** [comment]
Hi @lzhengchun. Depending on your code organization, there may be way to reorder weights only on the first inference, cache the reordered weights, and then reuse them.

The DNNL usage model is to create primitives once and then reuse them. So in this sense you can cache the reordered weights and only pass new inputs on each inference.

If this is not possible, you can rely on the fact that for the given primitive parameters (shapes, attributes, etc), DNNL would create a primitive (descriptor) that would require the same layouts. So if you cache reordered weights (probably as a `dnnl::memory`), you would be able to avoid reorders still even with all the checks.

Hope this helps.

**Comment by lzhengchun** [comment]
Thanks @rsdubtso that should help. I will give it a try to see how much I can save

**Comment by lzhengchun** [comment]
@rsdubtso FYI, I load the weights and execute all weight reorder before doing actually inference. So, I do not need to run those reorder (cached) in inference. Probably because my model is not very big (1M weights) and reorder weights does not take much time, I did not see any noticeable performance improvement .

**Comment by rsdubtso** [comment]
This makes sense. You can check [verbose mode](https://intel.github.io/mkl-dnn/dev_guide_verbose.html) output from both versions to validate this assumption.

**Comment by lzhengchun** [comment]
Thanks @rsdubtso

**Comment by lzhengchun** [comment]
> Hi @lzhengchun,
> 
> The problem seems to happen here:
> 
> ```c++
>     // set all weights to 1 and bias to zero for debug
>     for (int i = 0; i < net_args.size(); i++){
>         if(net_args[i].find(DNNL_ARG_WEIGHTS) == net_args[i].end()){
>             continue;
>         }
>         ...
>     }
> ```
> 
> Here, what happens is that you initialize the buffers with `1.f` for **already reordered** weights. So before you start the program you have the following picture:
> 
> ```
> user_src(nchw)     : initialized
> user_weights(oihw) : uninitialized      !!! <-- should be initialized instead of conv_weights
> conv_weights(any)  : initialized with 1 !!! <-- could be left uninitialized
> ...
> ```
> 
> But then at run time there is a reorder: `user_weights --> conv_weights`, which overrides the ones you put with whatever garbage was in uninitialized user_weights. Please try change the program so that it initialize the original weights, then reorder will put `1.f` appropriately to the conv_weights.
> 
> Also, I am not sure I understand why the code works for the case where there is only one convolution. To me the problem should appear whenever the weights reorder is called (and that should happen for one layer as well). Moreover, when I tried to reproduce the problem on my side, I got incorrect results for one layer too. So it is really strange that for you it works. Well, anyways, the program should be fixed as I mentioned above.
> 
> Few notes:
> 
> 1. The program works for `format::tag = nchw` because in this case the weights would be in `oihw` format, and no reorder is required. Hence, the results are correct.
> 2. I see you want to run conv + relu. In the code you share with us ReLU is a separate primitive. Please note that it is beneficial to **fuse** it with the convolution computations. [Here](https://intel.github.io/mkl-dnn/dev_guide_attributes_post_ops.html) is the relevant part in the documentation. Also, there is an example that shows the benefits of fusing: [performance_profiling_cpp](https://intel.github.io/mkl-dnn/performance_profiling_cpp.html)
>    I am not sure I understa

Hello @emfomenk 
Does deconv support **fuse**? 

I got erros as follows when fuse relu with decon, it works well with conv and I use the same  way to fuse relu and deconv
```
status: 3
message: could not create a primitive descriptor iterator
```
